// @async-generators/from-emitter/dist/commonjs/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
if (Symbol["asyncIterator"] === undefined)
    (Symbol["asyncIterator"]) = Symbol.for("asyncIterator");
const subject_1 = require("@async-generators/subject");
function default_1(emitter, opts) {
    let _opts = Object.assign({
        onNext: "next",
        onError: "error",
        onDone: "done",
        selectNext: (x) => x,
        selectError: (x) => x,
        dispose: () => { }
    }, opts);
    let subject = subject_1.default();
    let disposed = false;
    const nextListener = (...args) => subject.next(_opts.selectNext(...args));
    const errorListener = (...args) => {
        if (disposed)
            return;
        subject.error(_opts.selectError(...args));
        dispose();
    };
    const doneListener = () => {
        if (disposed)
            return;
        subject.done();
        dispose();
    };
    const dispose = () => {
        emitter.removeListener(_opts.onNext, nextListener);
        emitter.removeListener(_opts.onError, errorListener);
        emitter.removeListener(_opts.onDone, doneListener);
        _opts.dispose();
        disposed = true;
        ;
    };
    emitter.addListener(_opts.onNext, nextListener);
    emitter.addListener(_opts.onError, errorListener);
    emitter.addListener(_opts.onDone, doneListener);
    subject.on("disposed", () => dispose());
    return subject;
}
exports.default = default_1;
